### ISSUE TITLE
Typescript version / definition files

### ISSUE THREAD

**Issue by alvaromartmart** [openseadragon/openseadragon#1628]
Hi,
I was wondering if there are any plans to eventually release a typescript version of OpenSeadragon, or at least release a .d.ts definition file.
I've found a few mentions to this topic in #456 but it appears no one is working on it.

In case anyone is interested, I started my own definition file, which you can find in this repo:
https://github.com/alvaromartmart/types-openseadragon
Bear in mind it is not complete and is based on the official documentation of v. 2.4.0 of OpenseaDragon.
If anyone is interested in collaborating, please let me know!

Whenever it is stable I would like to publish it to the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) repo or as an npm package so anyone can install it as `@types/openseadragon`

**Comment by iangilman** [comment]
Sweet! 

I don't foresee us switching to TypeScript, or creating a parallel version in TypeScript, but if it's possible to have a sidecar file that gives optional TypeScript support to OpenSeadragon, that would be lovely!

The one you've built so far already looks pretty comprehensive. Sounds like publishing to DefinitelyTyped when you're ready is a great plan. Let me know if there's anything that would be helpful on the OSD side.  For instance I imagine it would be good to reference in the docs somewhere.

**Comment by msalsbery** [comment]
@alvaromartmart 

Excellent! This just recently got added to my need-to-do-soon list, so it's great to see someone already started tackling it!

If I find anything I can contribute then I'll definitely be interested in collaborating

**Comment by IdanCo** [comment]
So helpful!!! Thank you!

**Comment by peterlazzarino** [comment]
@alvaromartmart are you still looking for help on this?

**Comment by alvaromartmart** [comment]
@peterlazzarino I am!
While the current state of the definition file suits my needs, there are some tasks I would like to work on at some point:
> Disclaimer: As I stated on the OP, I wrote the definition file based on the official documentation, mostly. At some points I went to the code to better understand some things that were not clear to me, but 95% was me *trusting* that the documentation is up to date with the implementation.
- [ ] Reviewing the current definitions. Maybe not thoroughly, but I don't use the whole range of OpenSeadragon features, so there are things I might have missed:
    - [ ] There are a lot of methods/constructors with optional parameters, which in my head translates to several function signatures. I tried to include as many supported signatures in the definition file, but some may be missing.
    - [ ] I don't consider myself the greatest js developer in the world, so feedback is welcome. For instance, I'm not always sure when to use an interface over a type aliases. 
- [ ] There are no tests. I have not investigated yet the proper way to write ts definition file tests.
- [ ] There is just one big declaration file, it may be a good idea to spread it between several, feature-related files, and export them all together.
- [ ] OSD plugins support: I have not used any plugins yet and have not looked into how they are implemented, so I am not sure how easy it will be to include plugin definitions too.

Looking further down the road:
* I think including the code documentation in the definition file would be great, as we would be getting intellisense in VSCode and others, making it much easier to work with OpenSeadragon.
* Publishing the definition file to DefinitelyTyped, or including it on the OpenSeadragon repo itself.

So the TL; DR version is: anyone willing to help with this stuff, feel free to contact me, I'm open for collaborators ;)

**Comment by MarkusKern** [comment]
Very helpful! Could you please give some brief code example for usage?

**Comment by MarkusKern** [comment]
...more specifically, just the basics for gettings started, import statement and contructor basically, I can't seem to get it to work

**Comment by alvaromartmart** [comment]
@markmatney yes, I should probably make some working example when I have the time... just to get you started, if you use npm you can:
```
npm i --save openseadragon
npm i --save-dev alvaromartmart/types-openseadragon
```
Then you should be able to import stuff in your code like this:
```html
<div id="osd-div" style="width: 400px; height: 400px"></div>
```
```typescript
import { Viewer } from 'openseadragon';
const viewer = new Viewer({
      id: 'osd-div',
      prefixUrl: 'https://openseadragon.github.io/openseadragon/images/',
      tileSources:
        {
          Image: {
            xmlns: 'http://schemas.microsoft.com/deepzoom/2008',
            Url: '//openseadragon.github.io/example-images/duomo/duomo_files/',
            Format: 'jpg',
            Overlap: '2',
            TileSize: '256',
            Size: {
              Width:  '13920',
              Height: '10200'
            }
          }
        }
    });
```

**Comment by kristinruben** [comment]
@alvaromartmart I'm interested in collaborating to get this over the finish line into a [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) PR! 🎉

**Comment by alvaromartmart** [comment]
@kristinruben sure! Actually, I was thinking on preparing the PR soon (maybe in the next couple of days), I'll give you the heads up when I get to it.
I will probably need some help writing some tests as recommended by the `DefinitelyTyped` guidelines.

**Comment by kristinruben** [comment]
@alvaromartmart awesome, would love to help!

**Comment by alvaromartmart** [comment]
@kristinruben I've created an [`openseadragon` branch in my DefinitelyTyped fork](https://github.com/alvaromartmart/DefinitelyTyped/tree/openseadragon) with the intention of preparing the PR from there.

It's already passing the linting enforced by `DefinitelyTyped`
```
cd types/openseadragon
npm run lint openseadragon
```
and can be compiled
```
tsc
```

So it's almost ready to go, the only thing is, the test file is almost empty. If you can come up with a couple of tests -keep in mind that they're not ran, only type-checked- feel free to add them to `openseadragon-tests.ts`.

**Comment by kristinruben** [comment]
@alvaromartmart I have a local branch `knr/openseadragon` that I'm ready to push. Just need push access if that's okay!

I've linted the `tsconfig.json` to remove the comments (my linter was complaining?) and added the following to `openseadragon-tests.ts`. Based on the other modules in Definitely Typed, this seemed sufficient, but happy to add more if that's preferable:

```
import Viewer, { Drawer, Viewport, MouseTracker } from 'openseadragon';

const viewer = new Viewer({ id: "viewerid" });

const viewport = new Viewport({ margins: {} });

const element = new Element();

const drawer = new Drawer({ viewer, viewport, element });

const mouseTracker = new MouseTracker({ element });
```

**Comment by alvaromartmart** [comment]
Thanks @kristinruben, I just gave you collaboration permissions.

**Comment by JoFrMueller** [comment]
Hey, thanks for you efforts! How does it currently look regarding types?

I'm manually creating some of the most important interfaces right now and would be super happy if I could get rid of them at some point.

**Comment by alvaromartmart** [comment]
I actually made a [PR to DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41279) a while ago so that hopefully one day we can install the types using
`
npm install @types/openseadragon
`

Unfortunately my PR seems to be stuck waiting for the first reviewer to check it again, and I have not been able to catch his/her attention.

@iangilman, as a library maintainer, maybe you could drop by the PR comments and say something? 😅

**Comment by JoFrMueller** [comment]
Thanks for the quick reply!

I read the comments inside of the PR you mentioned. As @RyanCavanaugh still gets mentioned as "was requested for review", it seems not correct. I don't know, but perhaps the PR state of "Changes requested" prevents the initial mentioned reviewer to get informed/have a look at it again. 5 weeks since your last adjustment seem to be quite a long time for a review. Your initial pull request was reviewed within 3 days. So I guess somebody wasn't notified and as the state of the PR is "Changes requested", they still wait for changes. So it _could be_ a good old deadlock...

However, amazing work! I just hope they move on with this.

**Comment by JoFrMueller** [comment]
I just realized they merged your pull request. So sweet! Do you know how this will progress from here? How will the types enter [NPM](https://www.npmjs.com/search?q=%40types%2Fopenseadragon)?

**Comment by alvaromartmart** [comment]
It's supposedly to be done automatically by DefinitelyTyped, I think. But I just checked and it's not yet listed in the npm registry... no idea why.

**Comment by alvaromartmart** [comment]
Ok, i'ts already been published: https://www.npmjs.com/package/@types/openseadragon

It still doesn't appear on search results, though. I guess they have some sort of cache.

**Comment by JoFrMueller** [comment]
Super nice!

Thank you very much for your amazing work and all the instant responses!

**Comment by iangilman** [comment]
@alvaromartmart yes, thank you for making this happen!

Should we list information about the types somewhere on the OpenSeadragon site so people who need them can more easily find them?

Also, what happens in the future when new features are added to OpenSeadragon… Should there be a process for updating the types?

**Comment by iangilman** [comment]
@alvaromartmart thank you for making a PR with the installation instructions; I'll get that deployed soon.

Any thoughts on how we should maintain the types going forward?

I'm working with @Dan503 on an OpenSeadragon-based comic viewer using TypeScript, and he had a few bits he needed to add to our local copy of the definitions. You can see those changes here:

https://github.com/pixfabrik/driftory-viewer/pull/5/commits/350b5560a363b76707ceb0c273bc04b134702dbe

Does that seem like a reasonable change to make to the "official" version? How do we go about that? A new PR on https://github.com/DefinitelyTyped/DefinitelyTyped?

**Comment by Dan503** [comment]
> Does that seem like a reasonable change to make to the "official" version?

The exact change I made was more of a quick hack to just do the bare minimum to get my copy working so it could probably be better. I don't have enough seadragon or typescript knowledge to do a quality fix. The types in the official version should get fixed though.

> How do we go about that? A new PR

Personally I think it would be easier to just manually maintain a local `.d.ts` file as part of this repository rather than make PRs against the `@types` library. 

It would automatically get downloaded when users npm install OpenSeadragon. 

Also, if you ever make changes to the API, updating the types is as simple as updating the local file as part of the API changes. No need to make a PR against a third party repository that you don't have control over.

**Comment by iangilman** [comment]
> Personally I think it would be easier to just manually maintain a local .d.ts file as part of this repository rather than make PRs against the @types library.

That does sound appealing! I don't know enough about the TypeScript ecosystem to know if that's a common scenario. @alvaromartmart any thoughts on that?

**Comment by alvaromartmart** [comment]
Yes, I think that would be much better than having to keep the types at DefinitelyTyped in sync (specially considering that OpenSeadragon releases patch versions, and DefinitelyTyped [only considers minor versions](https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library) in order to determine which types correspond to the library version ).

**Comment by iangilman** [comment]
@alvaromartmart (and anyone else here interested) maybe it's time to revisit moving the TypeScript definitions into this repository? With the recent 5.0.0 release, there are new APIs to provide types for ([for example](https://github.com/openseadragon/openseadragon/issues/2569#issuecomment-2292289295)), and at any rate, DefinitelyTyped versions have already drifted out of sync.

**Comment by Dan503** [comment]
Be aware that the DefinitelyTyped library was built for the purpose of giving developers a way to define types for old JS libraries that have either been abandoned or the maintainers of the library have no interest in adding TS types to the list of their responsibilities.

If the maintainers of this library are happy to add maintaining TS types to the list of their responsibilities, then it makes the most sense for the `.d.ts` files to be added directly to this repo. The maintainers of this library should not maintain the types for their own library through DefinitelyTyped.

**Comment by msalsbery** [comment]
I have my own fork of the definitelytyped definitions that I have fixed tons of incorrect definitions and added lots of the new stuff. I use it in my projects, but it turned into such a big task I haven't been able to finish and publish it to definitely typed yet

**Comment by msalsbery** [comment]
Also, as a maintainer, I am very interested in maintaining this 😄

**Comment by alvaromartmart** [comment]
It's been a while now since I last used Openseadragon and honestly I don't have the time to contribute to the project nowadays but just wanted to give my 👍🏻 to moving away from DefinitelyTyped and include ts definitions in this repo, I think it makes total sense and will ensure better DX and avoid the challenges of keeping versions in sync. My only concern is that I never got to find a proper way of supporting plugins as I mentioned [here](https://github.com/openseadragon/openseadragon/issues/1628#issuecomment-732132094) but you folks will probably figure it out as you're more familiar with the codebase 🙂 .

**Comment by iangilman** [comment]
@Dan503 Thank you for the perspective... That makes a lot of sense. 

@msalsbery Awesome... I guess a good start would be for you to add that definition file here in a PR. It doesn't have to be perfect, just better than what's on DefinitelyTyped. Then we can continue to refine it. 

@alvaromartmart Totally understood on your availability. Thank you for getting the ball rolling in the first place! What's your concern with plugins? I guess I'm not sure how they would be affected by type definitions.

**Comment by iangilman** [comment]
Just for reference, here is the DefinitelyTyped OSD definition file: 

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/openseadragon/index.d.ts

**Comment by iangilman** [comment]
BTW, does anyone know any good tools for evaluating how well the definition file matches the actual code? Going forward, how can we validate that we are keeping the definitions in sync with the code?

**Comment by iangilman** [comment]
> @iangilman I just went through the whole thread 🧵 linked here, thanks for sharing it BTW..
> 
> I would love to try it, at least bringing the types definitions to this repo as that would significantly help everyone in maintaining it and to make contributions to it.. (as I won't be able to contribute consistently due to my other commitments) 😃
> 
> And if you could also share steps that one should do or a guide on how to run OSD build locally, that would be super helpful

@pranavgoel29 Moving the conversation here since this is where it's relevant. Thank you for working on this! 

Here's how to run locally: 

https://github.com/openseadragon/openseadragon/blob/master/CONTRIBUTING.md

**Comment by cmahnke** [comment]
See also:
* https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/72428
* https://github.com/DefinitelyTyped/DefinitelyTyped/pull/72433

**Comment by iangilman** [comment]
@cmahnke You should be able to add a new `openseadragon.d.ts` to your project and just include the new bits. For example: 

```
declare namespace OpenSeadragon {
  interface Options {
    drawer?: 'webgl' | 'canvas' | 'html';
  }
}
```

**Comment by cmahnke** [comment]
@iangilman Thanks, already figured it out by now. But there is a questionable design decision on the TS side (https://github.com/microsoft/TypeScript/issues/36146) that makes it impossible to update existing definitions when a type change is needed. For example when adding `Element` to the various `*Button` options of `Viewer`...

**Comment by iangilman** [comment]
@cmahnke Cool. This is why we're talking about moving the types into OSD directly. 

@pranavgoel29 Are you still interested in working on that?

**Comment by pranavgoel29** [comment]
> @cmahnke Cool. This is why we're talking about moving the types into OSD directly. 
> 
> @pranavgoel29 Are you still interested in working on that?

Yes, sorry I had some busy weeks recently, I am hoping to start on this soon..

**Comment by iangilman** [comment]
@pranavgoel29 Thanks for the update... whenever works for you 😊

**Comment by iangilman** [comment]
@pranavgoel29 Are you still interested in working on this?

**Comment by rssaini01** [comment]
I’ve started working on updating the definition files according to the latest [OSD 5 docs](https://openseadragon.github.io/docs/) in [DefinitelyTyped/openseadragon](https://github.com/rssaini01/DefinitelyTyped/tree/master/types/openseadragon). Most likely, it will be completed in the coming weeks, and I’ll raise a PR to the DT package.

@iangilman If you’re planning to include types directly in OSD, I’d love to contribute. 😊

**Comment by rssaini01** [comment]
> I’ve started working on updating the definition files according to the latest [OSD 5 docs](https://openseadragon.github.io/docs/) in [DefinitelyTyped/openseadragon](https://github.com/rssaini01/DefinitelyTyped/tree/master/types/openseadragon). Most likely, it will be completed in the coming weeks, and I’ll raise a PR to the DT package.
> 
> [@iangilman](https://github.com/iangilman) If you’re planning to include types directly in OSD, I’d love to contribute. 😊

PR opened to add support for OpenSeadragon v5 types: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/73846

**Comment by iangilman** [comment]
@rssaini01 Thank you so much for doing this! Wonderful to have types brought up to date! 

And yes, I think it would be great to begin including the types directly in the OSD library. I don't think we should be moving the code base itself over to TypeScript (yet) but including the definition files in the OSD repository and package seems like an important next step. Let me know if you're still up for working on that!

**Comment by rssaini01** [comment]
@iangilman Yes, I’d like to take this up. Since we’ve already started a discussion here: [Discussion #2810](https://github.com/openseadragon/openseadragon/discussions/2810), I’ll share an update there in the coming days. The PR in the DefinitelyTyped package has already been merged and should include the latest v5 changes, as far as I can tell. 😊

**Comment by shirshamay** [comment]
Hello,  I would like to help i saw this issue is still open can I assign the mission please?

**Comment by iangilman** [comment]
@shirshamay Sorry, @rssaini01 is working on this. I've now officially assigned them, to avoid confusion. 

We welcome your help, though, if you want to work on another issue!

**Comment by shawshank-fork** [comment]
Hey, i would like to help, i there any still needed to be done? Do let me know.

**Comment by iangilman** [comment]
@shawshank-fork I think @rssaini01 has it under control, thanks!

In fact, I think at this point we can retire this issue. If we need more specific issues for various aspects of our TypeScript definitions, we can create them.
